Feature: Linux - Add native ALSA MIDI output support

drelbszoomer opened the pull request:
## Motivation / Problem / Description

This adds a proper/native Linux ALSA MIDI sequencer output driver to OpenTTD, as the native alternative on Linux platforms to the `extmidi` interface, which is a  bit kludgy and can't work with hardware MIDI devices.

I've tested this with a hardware MIDI device (Roland SC-55 over USB-MIDI) and it works pretty well for me, but feedback welcome.

This also adds a small 3rd party MIDI library for Standard MIDI file/event/timing parsing and correction, and (obviously) a `libasound` dependency if built on Linux.

To test, ensure you have `libasound-dev` or your Linux distro's equivalent installed before you build.

Supports the standard MIDI driver arguments, so if your MIDI device is on port 32 like mine is (UM-ONE USB):

```
$ aplaymidi -l                                                                                                                                                                                                                           
 Port    Client name                      Port name
 14:0    Midi Through                     Midi Through Port-0
 32:0    UM-ONE                           UM-ONE MIDI 1
``` 

You can just do

`./openttd -m "alsamidi:port=32"`

Debug logging for the driver is available with `-d driver2`.

## How is this different from the other MIDI options OpenTTD has that already work on Linux (extproc and Fluidsynth)

- `extproc` doesn't interface with the system MIDI bus at all, and just (somewhat awkwardly) works with a handful of MIDI apps by launching them as subprocesses and feeding them the filename. Only a few kinds of MIDI playback apps are supported, and the experience is not ideal. It is, for instance, possible on Linux to get Timidity++ to work with the `extproc` driver, where TM is invoked directly as a playback app, but the control, features, and experience are pretty crude/simplistic. This `ALSA`  driver would be a much better choice than `extproc` for Timidity (or any other software synth that runs on Linux) that can expose itself as a MIDI sink on the system MIDI bus.

- `FluidSynth` is a software midi synthesizer - it can't interface over MIDI ports with real devices. This `ALSA` driver is the Linux equivalent of the `win32/dmusic/cocoa` drivers, where MIDI events are sent thru the _system_ MIDI bus to whatever device is connected to the system. Fluidsynth just  renders MIDI files to PCM in-process  without interacting with the system MIDI bus or connected devices at all. It does a good job of that.

Having native MIDI support allows you to use either 

- the real hardware devices the music was originally composed for
- any other kind of software synth that can connect to the system MIDI bus (for instance, https://github.com/nukeykt/Nuked-SC55 would also work with this out of the box)
- This would also work with Timidity++ on Linux in daemon mode (another software synth that can expose itself as a system MIDI sink), and would be a better option than the existing `extproc` driver (which can only use Timidity++ by launching it as a direct process, not as system MIDI device) in that scenario as well.

Things like most DOOM engines, ScummVM etc support both Fluidsynth and native MIDI playback and it seems reasonable that OpenTTD should have this too.

## Limitations
There's probably strange MIDI packs out there that can break this, and I haven't tested the SYSEX stuff very much (none of OpenTTD's downloadable music packs seem to have SYSEX messages)


## Checklist for review

Some things are not automated, and forgotten often. This list is a reminder for the reviewers.
* The bug fix is important enough to be backported? (label: 'backport requested') -> NO
* This PR touches english.txt or translations? Check the [guidelines](https://github.com/OpenTTD/OpenTTD/blob/master/docs/eints.md) -> NO
* This PR affects the save game format? (label 'savegame upgrade') -> NO
* This PR affects the GS/AI API? (label 'needs review: Script API') -> NO
* This PR affects the NewGRF API? (label 'needs review: NewGRF') -> NO

drelbszoomer committed:
Feature: Add ALSA midi output support

PeterN commented:
This comment is not a request for changes, I'm just writing my thoughts here for others.

I'm not too keen on vendoring more libraries if it can be avoided. We do have custom midi file processors, sadly not yet written in a generic way. It's possible that that existing code could be unified and abstracted and used instead of bringing in more code.

drelbszoomer commented:
> This comment is not a request for changes, I'm just writing my thoughts here for others.
> 
> I'm not too keen on vendoring more libraries if it can be avoided. We do have custom midi file processors, sadly not yet written in a generic way. It's possible that that existing code could be unified and abstracted and used instead of bringing in more code.

Yes, I sympathize with that. And it occurred to me when adding this. I do think it makes sense to externalize MIDI file handling because they come in all shapes and sizes and writing code to handle all of them correctly in all edge cases can be a bit tricky.

TrueBrain commented:
Out of curiosity, have you first tried FluidSynth? As we already ship that in our official binaries, and as far as I am aware it has dynamic linking against ALSA: so that should, in theory, do something very similar.

I am not familiar with how MIDI is handled, so I might be saying very silly :) mostly curious why FluidSynth isn't enough (mostly as I thought it was meant to solve this issue :D )

drelbszoomer commented:
> Out of curiosity, have you first tried FluidSynth? As we already ship that in our official binaries, and as far as I am aware it has dynamic linking against ALSA: so that should, in theory, do something very similar.

FluidSynth is a software midi synthesizer - it can't interface over MIDI ports with real devices.

This driver is the Linux equivalent of the win32/dmusic/cocoa drivers, where MIDI events are sent thru the system MIDI bus to whatever device is connected to the system. Fluidsynth just  renders MIDI files to PCM in-process  without interacting with the system MIDI bus or connected devices at all.

PeterN commented:
Ah, sorry I didn't address that issue because I do have real hardware synths that I'd like to use without extmidi and it was obvious to me what this is about :-)

nielsmh commented:
On the use of a 3rd party MIDI file parser, do consider that OpenTTD still needs its own MIDI data handling for the GM.DAT original Transport Tycoon Deluxe for DOS music data format. The data loader I wrote back then abstracts data loading from standard MIDI files and the "MPS-MIDI" format.

PeterN commented:
> On the use of a 3rd party MIDI file parser, do consider that OpenTTD still needs its own MIDI data handling for the GM.DAT original Transport Tycoon Deluxe for DOS music data format. The data loader I wrote back then abstracts data loading from standard MIDI files and the "MPS-MIDI" format.

Fancy abstracting your code out to make it reusable here instead of this 3rd party lib? I've looked at doing so a few times (for an ALSA MIDI driver, no less) but decided it's a side-quest I can do without for now. :-)

nielsmh commented on src/music/alsamidi.cpp:
```diff
@@ -0,0 +1,432 @@
+
+#include "../safeguards.h"
+
+/** Factory for BeOS' midi player. */
```
Not BeOS :)

nielsmh commented on src/music/alsamidi.cpp:
```diff
@@ -0,0 +1,432 @@
+				ev[1],  // Note number
+				ev[2]   // Velocity
+			);
+		}
+		else if (ev.isNoteOff()) {
```
There shouldn't be a line break here.

nielsmh commented on src/music/alsamidi.cpp:
```diff
@@ -0,0 +1,432 @@
+		snd_seq_ev_set_sysex(&seqev, complete_message.size(), complete_message.data());
+}
+
+void MusicDriver_AlsaMidi::SendEvent(smf::MidiEvent& ev) {
```
Should have a line break before the opening brace of a function body.

nielsmh commented on src/music/alsamidi.cpp:
```diff
@@ -0,0 +1,432 @@
+		Debug(driver, 2, "ALSA MIDI: stopping current queue on reset");
+		// Drop anything still in the queue, this is a disruptive reset.
+		snd_seq_drop_output(this->seq);
+		// Reset all controllers to their defaults
```
In the Win32 MIDI driver this is handled by sending a GM System Reset sysex, which I've generally found to work more reliably.

drelbszoomer replied:
Yeah - I struggled a bit to get a reliably working reset, and settled on this because I read somewhere that not all devices support GM sys reset.

I'll give that a shot tho, and if the other drivers do it, that's good enough for me.

drelbszoomer replied:
Yep, that works well enough  and is simpler, thanks!

nielsmh commented on src/music/alsamidi.cpp:
```diff
@@ -0,0 +1,432 @@
+
+void MusicDriver_AlsaMidi::SetVolume([[maybe_unused]]uint8_t vol)
+{
+	// TODO no-op, output device has its own volume control
```
The volume control in the sound/music window should do something. The Win32 driver handles this the same way as Transport Tycoon did, by setting the per-channel volumes and scaling all channel volume controller messages. However that does mean that changing volume via the in-game UI can cause a bunch of messages to flow down the cable, especially if the user is swinging the volume slider around wildly.

drelbszoomer replied:
Yeah - I considered doing it as channel vol. I personally would probably never use the in-game volume control. I'll give it a shot, tho.

drelbszoomer replied:
All right, I implemented this. I'm using a relatively simplistic debounce to avoid the scenario you mentioned, which works pretty well on my end.

nielsmh reviewed (commented):
Make sure to review the coding style guidelines.

nielsmh commented:
> Fancy abstracting your code out to make it reusable here instead of this 3rd party lib? I've looked at doing so a few times (for an ALSA MIDI driver, no less) but decided it's a side-quest I can do without for now. :-)

To be quite honest, I misremembered what I'd done and thought the data was already in a more abstracted format, rather than just "barely-chewed chunks of raw MIDI" - yeah a lot of the more complex handling does get done in the Win32 (and DMusic) drivers rather than the `MidiFile` class.

drelbszoomer commented:
UPDATED:
  - Switch to using `pollfds` instead of busy waits, which is more idiomatic/less expensive for ALSA/Linux.
  - Implemented volume control with debounce.
  - Align with CODINGSTYLE.md (I might have missed some things)
  
  @PeterN / @nielsmh PTAL when you have a chance and point out anything I've missed or that needs to be done still - thank you!
